CMakeLists.txt

```diff
@@ -228,22 +228,33 @@ endif()
 # GNUInstallDirs.
 include(GNUInstallDirs)
 
-find_package(nlohmann_json QUIET)
-if(NOT nlohmann_json_FOUND)
-  message("Using local nlohmann::json from submodule")
-  set(JSON_BuildTests
-      OFF
-      CACHE INTERNAL "")
-  set(JSON_Install
-      OFF
-      CACHE INTERNAL "")
-  # This option allows to link nlohmann_json::nlohmann_json target
-  add_subdirectory(${PROJECT_SOURCE_DIR}/third_party/nlohmann-json)
-  # This option allows to add header to include directories
-  include_directories(
-    ${PROJECT_SOURCE_DIR}/third_party/nlohmann-json/single_include)
-else()
-  message("Using external nlohmann::json")
+if(NOT WITH_API_ONLY)
```

Comment:
@lidavidm - I hope that should fix the build for API-only build.

Reply:
I can confirm this fixes the API-only build, thanks!

It would be nice if we could still build the rest of the library that doesn't require json without this dependency, though (e.g. the SDK), though I can understand if that would get complicated.

Reply:
@lidavidm - it happens that in the majority of cases (`80% scenarios` I'd say), we do actually require `nlohmann/json.hpp`, e.g. for ETW on Windows, for zPages (internal monitor of SDK statistics), for Zipkin exporter, as well as for OTLP-HTTP exporter. Plus for all tests, and generally for examples too. Since we got you covered for the case when the build is done only for API and nothing else, I think it's gonna be too much of a hassle to make json optional.... as you noticed, it gets complicated - given that most components require it. And we give two alternatives - either consume json library from user-provided deps. Or use the one we cloned from recursive submodules, both no work cross-plat.

We can probably consider further improvements in a follow-up PR, but as low priority. Since all immediate scenarios have been addressed.

Reply:
I agree on having further cleanup in a follow-up PR. I think the most common usage scenario would be instrumentation libraries devs using otel-cpp api, and application developers using otel-gRPC exporter. And both these scenarios don't depend on JSON and so shouldn't fail. Though I do see the hassle involved here to work it cross-platform.
 
Thanks for fixing the API only build :)

Reply:
@lalitb - My worry is that OTLP-HTTP exporter may also be common. This one does require JSON. Same as Zipkin. Also I think the zPages project is kinda common across all languages. The idea is we use zPages for debugging OpenTelemetry in-proc.. that build flavor of final SDK dll / shared object bit, with zPages in it, would require json.hpp

While I agree we may not need it for OTLP-gRPC only build without zPages, in most other configurations we end up needing json.hpp anyways.

Reply:
I totally understand. I think the biggest thing is just that the release tarballs don't include the submodules, but that's a Github issue.